In [27]:
import os
import sys
import numpy as np
import pylab as pl

%pylab inline



Populating the interactive namespace from numpy and matplotlib


NULL HYPOTHESIS: the % of former prisoners employed after release is the same or lower for candidates who participated in the program as for the control group, significance level p=0.05
 $H_0: P_0 - P_1 \geq$ 0

 $H_a: P_0 - P_1 $< 0

 $\alpha$ = 0.05

this is a TEST OF PROPORTIONS. we use the Binomial distribution since it is a yes/no (bernulli) test for each subject: the former inmate was or was not ever employed in a CEO transitional job (second row in the table above):

 $P_0=0.035, P_1=0.701$

In [28]:
# first thing: choose your significance threshold
alpha=0.05



# we like fractions better then percentages. as a rule of thumb, either use fractions or counts. 
# 0 for control group, 1 for program group

P_0 = 3.5 * 0.01 
P_1 = 70.1 * 0.01

if P_0 - P_1 >= 0:
    # we are done
    print ("the Null holds")
else:
    print ("we must assess the statistical significance")

#sample sizes 

n_0 = 409
n_1 = 564
    
#lets get the counts by multiplying by the sample size

Nt_0 = P_0 * n_0
Nt_1 = P_1 * n_1

we must assess the statistical significance


WHICH TEST?
samples, categorical data

TWO OPTIONS z test, or chi-square test.

## START WITH Z TEST
#### the z test compares the stanrard deviation of the expected distribution and the observed result. it tells you literally how many standard deviations from the tail an observation is, under the assumption of normality

must define the sample standard deviation (standard error)

In [29]:
#define the sample proportion first
sp = (P_0 * n_0 + P_1 * n_1) / (n_1 + n_0)
print (sp)

0.4210472764645426


standard deviation of the sampling distribution the distribution is Binomial, the binomial stdev is

(see a proof here!: http://stats.stackexchange.com/questions/29641/standard-error-for-the-mean-of-a-sample-of-binomial-random-variables!):

$\sqrt{\frac{p(1 - p)}{n}}$

for 2 samples this becomes

$\sqrt{ \frac{ \hat{p}(1 - \hat{p})} {n0} + \frac{ \hat{p}(1 - \hat{p})} {n1} }$

cfr: page 138 of Statistics in a Nutshell, eq. 5.12 and here http://stattrek.com/hypothesis-test/difference-in-proportions.aspx?Tutorial=AP

 $\hat{p} =\frac{p_0  n_0 + p_1  n_1}{n_0+n_1}$

Note that in the online version of Statistics in a Nutshell, $\bar{x}$ should be  $\hat{p}$!!

In [30]:
# I am goonna create a little one line function to calculate the standard error
# and to calculate p

p = lambda p0, p1, n0, n1: (p0 * n0 + p1 * n1) / (n0 + n1)
#standard error
se = lambda p, n0, n1: np.sqrt(p * (1 - p) * (1.0 / n0 + 1.0 / n1))

z score: how many standard deviation away from the population parameter is my statistic?

$z=\frac{P_1-P_0}{\sigma}$

In [31]:
zscore = lambda p0, p1, s : (p0 - p1) / s
z_2y = zscore(P_1, P_0, se(p(P_0, P_1, n_0, n_1), n_0, n_1))
print (z_2y)


20.7697865408


note that using p0-p1 or p1-p0 at the numerator is equivalent because the standardizes normal value of z has mean 0 (see image below) so that we can use the absolute value of the z score, or equivalently look for $P[Z&lt;z]$ if z is positive, and $P[Z&gt;z]$ if z is negative.

if $p&lt;\alpha$ : reject H0 **
IMPORTANT!! note that this P in the bottom line of the table is not the p-value, but

p-value = 1-P

In [32]:
## p-value for employment after 2 years: 
## since the largest number we read off the table for is (way) smaller 
## than the value for our statistic 
## our p-value will be smaller than it would be if calculated using 
## (e.g.) .9998 (and in fact using 1.0000 which is the largest number 
## in the table). Using 0.9998 is a **conservative** approach. 

p_2y = 1 - 0.9984


def report_result(p,a):
    print ('is the p value ' + 
           '{0:.2f} smaller than the critical value {1:.2f}?'.format(p,a))
    if p < a:
        print ("YES!")
    else: 
        print ("NO!")
    
    print ('the Null hypothesis is {}'.format(\
                            'rejected' if p < a  else 'not rejected') )

    
report_result(p_2y, alpha)

is the p value 0.00 smaller than the critical value 0.05?
YES!
the Null hypothesis is rejected


### Results of our test for the overall employment data:
Since we have rejected the null hypothesis, we are saying that former prisoners who participate in the CEO program have a higher rate of employment than those who do not participate in the program and that this difference is real, statistically significant, and likely not due to chance.

### what if we used the values for where the former inmate was or was not "Convicted of a felony" (row 10) in the Recidivism (Years 1-3)?

### Null Hypothesis 

Of participants who experienced Recidivism within 3 years and were convicted of a crime, those who completed the CEO employment program are just as or more likely to be convicted of a felony as those who didn't complete CEO employment training.

H_0: P_0 - P_1 <= 0

H_a: P_0 - P_1  > 0

P_0 = rate of felony convictions (control)

P_1= rate of felony convictions (test) 

alpha=0.05

In [33]:
alpha=0.05

# p_0 = .117
# p_1 = .10

P_0 = 11.7 * 0.01
P_1 = 10.0 * 0.01

### if P_0 - P_1 <= 0:
#    print ("Null holds")
# else:
 #   print ("we must asses the statistical signficance")
    
N_0 = 409
N_1 = 558

Nt_0 = P_0 * N_0
Nt_1 = P_1 * N_1

z_3y = zscore(P_1, P_0, se(p(P_0, P_1, N_0, N_1), N_0, N_1))
print(z_3y)

-0.844223197776


In [34]:
p_3y = 1 - 0.8023 #see standard stat table for p value

def report_result(p,a):
    print ('is the p value ' + 
           '{0:.2f} smaller than the critical value {1:.2f}?'.format(p,a))
    if p < a:
        print ("YES!")
    else: 
        print ("NO!")
    
    print ('the Null hypothesis is {}'.format(\
                            'rejected' if p < a  else 'not rejected') )

report_result(p_3y, alpha)



is the p value 0.20 smaller than the critical value 0.05?
NO!
the Null hypothesis is not rejected


The p value is not 0.20 smaller than our alpha(0.05) therefore we cannot reject our null hypothesis. People who completed the CEO education program are more or just as likely to be a felon as those who did not complete the CEO education program. 




|                 |     success         | failure|    |               
|-----------------|:-------------------:|:-------------------:|---------------------------|
| test sample     | number of successes in test    | number of failures in test    | number members of test sample |
| control sample  | number of successes in control | number of failures in control | number members of control sample| 
|                 | total successes                |  total failures               | number of all members         |

|employed in subsadized job |     employed          | not employed     |                   
|---------------------------|:---------------------:|------------------|---------------------------|
| test sample               | $0.701*564$           | $0.299*564$      | 564                       |
| control sample            | $0.035*409$           | $0.965*4.09$     | 409                       |
|                           |                       |                  |                           |
| total                     | 409.679               |  562.912         | 973                       |


In [35]:
# from evalChisq import evalChisq - AMBER CHEN HElPED ME WITH THIS, WITHOUT HER ID BE LOST 

def evalChisq(values):
    values = np.array(values)
    E = np.empty_like(values)
    for j in range(len(values[0])):
        for i in range(2):
            
            E[i][j] = ((values[i,:].sum() * values[:,j].sum()) / 
                        (values).sum())
    return ((values - E)**2 / E).sum()

In [36]:
Ntot = 973 # a + b + c + d = tot

sample_values = np.array([[0.701 * 564, 0.299 * 564], [0.0305 * 409, 0.965 * 409]])

print (evalChisq(sample_values))

436.223462575


This number must be compared to the chi sq distribution. You must calculate the number of degrees of freedom forthis experiment. Generally: DOF = Number of observations - number of Independent Variables

so here DOF = 1. Now you can look at the table below and draw conclusions about the rejection of the Null

# CHi-SQ table 

* fill in the contingency table below *

|convicted of a fellony     |     yes   | no        |                   
|---------------------------|-----------|-----------|----------------|
| test sample               |0.1*568  |  0.90*568         | 568               |
| control sample            |0.117*409           | 0.883*409          |       409         |
|                           |           |           |                |
| total                     | 104.653          | 872.347          |  977              |

432 is hella larger then 3.84

why am i mentioning 3.84?

how does the chi square statistic that you derived compare?

please state what that means in terms of your Null hypothesis in a markdown cell below!

In [37]:
chisq_felon = np.array([[0.1 * 568, 0.9 * 568], [0.117 * 409, 0.883 * 409]])
print (evalChisq(chisq_felon))

0.718493917505


3.84 is the number that corresponds to a degree of freedom of 1 and a probability of a larger value of x^2 in the chi-sq distribution table. 

436 is out of the range of the chisq distribution table meaning that our null hypothesis is very unlikely to ever be rejected based on our sample size.

The chi square stat of .718 should correspond with a degree of freedom and then a probability of a larger value of x^2. However, that does not correspond with a number on the chisq table so we will have to take a range (0.5 and .25). This would confirm that our null hypothesis can not be rejected and that the CEO employee program has no effect on the program/control felon groups. 
